include/merlin/core_kernels.cuh

```diff
@@ -1966,5 +1968,669 @@ __global__ void dump_kernel(const Table<K, V, M>* __restrict table,
   }
 }
 
+/* If founds[i] = true, read data from corresponding address of table_value_addrs and write to param_values;
+ * if founds[i] = false, write data from param_values to corresponding address of table_value_addrs.
+ * usually called by find_or_insert kernel.
+ */
+template <class V>
+void read_or_write_by_cpu(V** __restrict table_value_addrs, V* __restrict param_values,
+                  const int* __restrict offset, const bool* founds, size_t dim, int N,
+                  int n_worker = 16) {
+  std::vector<std::thread> thds;
+  if (n_worker < 1) n_worker = 1;
+
+  auto functor = [=](V** __restrict table_value_addrs, V* __restrict param_values,
+                    const int* __restrict offset, int handled_size,
+                    int trunk_size) -> void {
+    for (int i = handled_size; i < handled_size + trunk_size; i++) {
+      if (founds[offset[i]] && table_value_addrs[i] != nullptr) {
+        memcpy(param_values + offset[i] * dim, table_value_addrs[i], sizeof(V) * dim);
+      }
+      else {
+        memcpy(table_value_addrs[i], param_values + offset[i] * dim, sizeof(V) * dim);
+      }
+    }
+  };
+
+  size_t trunk_size = N / n_worker;
```

Comment:
If `N < n_worker` is true, trunk_size will be `0`.

Reply:
OK, I will check before launch a thread to avoid its overhead

Reply:
And try to make every `worker`'s workload balanced